### PR TITLE
ROX-23956, ROX-17355: re-index images

### PR DESCRIPTION
**Pull request by RTann**
### Description

Re-index images. This is based on https://github.com/stackrox/stackrox/pull/12613.

### User-facing documentation

- [x] CHANGELOG is updated
- [x] documentation PR is not needed

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [x] CI results are inspected

#### Automated testing

- [x] added unit tests

#### How I validated my change

1. Deployed StackRox 4.5.4 to an OpenShift cluster
2. Waited a sufficient amount of time for Scanner V4 to initialize
3. Scanned three images: `alpine:3.19`, `alpine:3.20`, `registry.access.redhat.com/ubi8:latest`
  Note: The RHEL image has one fixable vulnerability - [RHSA-2024:7848](https://access.redhat.com/errata/RHSA-2024:7848) (image SHA `sha256:d85bc4ef93c70be9396f7079262da29ae1f2b69eeb999aae57a5ef8301576dca`) but 4.5.4 won't find it within the first 24 hours of its lifetime...
4. Sanity check DB:
```
# SELECT COUNT(*) FROM indexreport;
 count 
-------
     3
(1 row)
```
5. Upgrade Scanner V4 to this PR (4.7.x-108-g3ae55f519f)
6. Upon restart, check Indexer logs:
```
{"level":"info","host":"scanner-v4-indexer-5dc745f88-l6hxz","component":"indexer/manifest/Manager.MigrateManifests","migrated_manifests":3,"time":"2024-11-11T22:49:58Z","message":"migrated missing manifest metadata"}
...
{"level":"info","host":"scanner-v4-indexer-5dc745f88-l6hxz","component":"datastore/postgres/indexerMetadataStore.GCManifests","deleted_manifests":3,"time":"2024-11-11T22:49:58Z","message":"deleted manifests"}
{"level":"info","host":"scanner-v4-indexer-5dc745f88-l6hxz","component":"indexer/manifest/Manager.Start","deleted_manifest_metadata":3,"time":"2024-11-11T22:49:58Z","message":"deleted expired manifest metadata"}
{"level":"info","host":"scanner-v4-indexer-5dc745f88-l6hxz","component":"indexer/manifest/Manager.Start","time":"2024-11-11T22:49:58Z","message":"next manifest metadata GC run will be in about 6h5m0s"}
```
7. Check the DB to find manifests have been deleted:
```
# SELECT COUNT(*) FROM indexreport;
 count 
-------
     0
(1 row)

# SELECT COUNT(*) FROM manifest_metadata;
 count 
-------
     0
(1 row)
```
8. Rescan the images and check the DB:
```
# SELECT * FROM manifest_metadata;
                                                               manifest_id                                                               |     expiration      
-----------------------------------------------------------------------------------------------------------------------------------------+---------------------
 sha512:23f78746cfa9194d6be42c80f355210697436bd04908ae7bb04f5307c6b391adfbb241b3477c2c55cfd67e02ee9ea105f7e505a8b755cb8273f2f60f3f876163 | 2024-12-02 01:19:46
 sha512:48032905aa1902b9eaab33db53641f0ad919c12e80e092b9b299e2b52dffdcbc2a18f1517e3c77334677a56a3fe345ade505a0103ce315255b158d50c598cf7b | 2024-11-20 09:07:04
 sha512:e98b94787cc5e79f66d225411301c4fa4fcf6b4a48334691eb7dd10be48d2e364b7eec0b37d7f9e65d233e7047ce728540b528b1b52475d5c341e70e78fee33c | 2024-11-19 06:38:27
(3 rows)
```
9. Mimic a manifest metadata update failure by deleting a row:
```
# DELETE FROM manifest_metadata WHERE manifest_id = 'sha512:23f78746cfa9194d6be42c80f355210697436bd04908ae7bb04f5307c6b391adfbb241b3477c2c55cfd67e02ee9ea105f7e505a8b755cb8273f2f60f3f876163';
DELETE 1
# SELECT * FROM manifest_metadata;
                                                               manifest_id                                                               |     expiration      
-----------------------------------------------------------------------------------------------------------------------------------------+---------------------
 sha512:48032905aa1902b9eaab33db53641f0ad919c12e80e092b9b299e2b52dffdcbc2a18f1517e3c77334677a56a3fe345ade505a0103ce315255b158d50c598cf7b | 2024-11-20 09:07:04
 sha512:e98b94787cc5e79f66d225411301c4fa4fcf6b4a48334691eb7dd10be48d2e364b7eec0b37d7f9e65d233e7047ce728540b528b1b52475d5c341e70e78fee33c | 2024-11-19 06:38:27
(2 rows)
```
10. Rescan all images and check the DB to find the two previously indexed images remain the same and the deleted manifest is re-indexed:
```
# SELECT * FROM manifest_metadata;
                                                               manifest_id                                                               |     expiration      
-----------------------------------------------------------------------------------------------------------------------------------------+---------------------
 sha512:48032905aa1902b9eaab33db53641f0ad919c12e80e092b9b299e2b52dffdcbc2a18f1517e3c77334677a56a3fe345ade505a0103ce315255b158d50c598cf7b | 2024-11-20 09:07:04
 sha512:e98b94787cc5e79f66d225411301c4fa4fcf6b4a48334691eb7dd10be48d2e364b7eec0b37d7f9e65d233e7047ce728540b528b1b52475d5c341e70e78fee33c | 2024-11-19 06:38:27
 sha512:23f78746cfa9194d6be42c80f355210697436bd04908ae7bb04f5307c6b391adfbb241b3477c2c55cfd67e02ee9ea105f7e505a8b755cb8273f2f60f3f876163 | 2024-11-30 08:08:34
(3 rows)
```
11. Delete another manifest and then restart one Indexer. This triggers a migration of the deleted manifest to the manifest_metadata table and then the manifest is immediately garbage collected:
```
{"level":"info","host":"scanner-v4-indexer-5dc745f88-9tk2v","component":"indexer/manifest/Manager.MigrateManifests","migrated_manifests":1,"time":"2024-11-11T23:08:38Z","message":"migrated missing manifest metadata"}
...
{"level":"info","host":"scanner-v4-indexer-5dc745f88-9tk2v","component":"datastore/postgres/indexerMetadataStore.GCManifests","deleted_manifests":1,"time":"2024-11-11T23:08:38Z","message":"deleted manifests"}
{"level":"info","host":"scanner-v4-indexer-5dc745f88-9tk2v","component":"indexer/manifest/Manager.Start","deleted_manifest_metadata":1,"time":"2024-11-11T23:08:38Z","message":"deleted expired manifest metadata"}
{"level":"info","host":"scanner-v4-indexer-5dc745f88-9tk2v","component":"indexer/manifest/Manager.Start","time":"2024-11-11T23:08:38Z","message":"next manifest metadata GC run will be in about 5h55m0s"}
```
12. Update the environment variables to ensure the minimums are used. Set `ROX_SCANNER_V4_MANIFEST_DELETE_DURATION`, `ROX_SCANNER_V4_MANIFEST_DELETE_INTERVAL_START`, and `ROX_SCANNER_V4_MANIFEST_GC_INTERVAL` each to `1s`:
```
{"level":"warn","host":"scanner-v4-indexer-79d854596c-f9hdz","component":"scanner/backend/indexer.NewIndexer","time":"2024-11-11T23:17:35Z","message":"configured manifest GC interval (1s) is too small: setting to 1h0m0s"}
{"level":"info","host":"scanner-v4-indexer-79d854596c-f9hdz","component":"indexer/manifest/Manager.MigrateManifests","migrated_manifests":0,"time":"2024-11-11T23:17:35Z","message":"migrated missing manifest metadata"}
{"level":"warn","host":"scanner-v4-indexer-79d854596c-f9hdz","component":"scanner/backend/indexer.NewIndexer","time":"2024-11-11T23:17:35Z","message":"configured manifest delete interval (1s) start is too small: setting to 1m0s"}
{"level":"warn","host":"scanner-v4-indexer-79d854596c-f9hdz","component":"scanner/backend/indexer.NewIndexer","time":"2024-11-11T23:17:35Z","message":"configured manifest delete interval (1s) duration is too small: setting to 2m0s"}
```
9. Updated each column of the table as such to be deleted within an hour:
```
# SELECT * FROM manifest_metadata;
                                                               manifest_id                                                               |     expiration      
-----------------------------------------------------------------------------------------------------------------------------------------+---------------------
 sha512:23f78746cfa9194d6be42c80f355210697436bd04908ae7bb04f5307c6b391adfbb241b3477c2c55cfd67e02ee9ea105f7e505a8b755cb8273f2f60f3f876163 | 2024-11-11 23:27:00
 sha512:48032905aa1902b9eaab33db53641f0ad919c12e80e092b9b299e2b52dffdcbc2a18f1517e3c77334677a56a3fe345ade505a0103ce315255b158d50c598cf7b | 2024-11-11 23:40:00
(2 rows)
```
14. Come back to eventually find these entries have all been deleted (note: these are "inactive" images which StackRox does not reprocess, so this is expected)
```
# SELECT * FROM manifest_metadata;
 manifest_id | expiration 
-------------+------------
(0 rows)

# SELECT * FROM indexreport;
 state | scan_result | manifest_id 
-------+-------------+-------------
(0 rows)

# SELECT * FROM manifest;
 hash | id 
------+----
(0 rows)
```
15. Rescan all the images (be sure to use the `-f` (force) flag)
16. Find all the images are there again:
```
# SELECT * FROM manifest_metadata;
                                                               manifest_id                                                               |     expiration      
-----------------------------------------------------------------------------------------------------------------------------------------+---------------------
 sha512:48032905aa1902b9eaab33db53641f0ad919c12e80e092b9b299e2b52dffdcbc2a18f1517e3c77334677a56a3fe345ade505a0103ce315255b158d50c598cf7b | 2024-11-12 01:01:09
 sha512:23f78746cfa9194d6be42c80f355210697436bd04908ae7bb04f5307c6b391adfbb241b3477c2c55cfd67e02ee9ea105f7e505a8b755cb8273f2f60f3f876163 | 2024-11-12 01:01:24
 sha512:e98b94787cc5e79f66d225411301c4fa4fcf6b4a48334691eb7dd10be48d2e364b7eec0b37d7f9e65d233e7047ce728540b528b1b52475d5c341e70e78fee33c | 2024-11-12 01:01:01
(3 rows)

# SELECT * FROM manifest;
                                                                  hash                                                                   | id 
-----------------------------------------------------------------------------------------------------------------------------------------+----
 sha512:48032905aa1902b9eaab33db53641f0ad919c12e80e092b9b299e2b52dffdcbc2a18f1517e3c77334677a56a3fe345ade505a0103ce315255b158d50c598cf7b | 10
 sha512:23f78746cfa9194d6be42c80f355210697436bd04908ae7bb04f5307c6b391adfbb241b3477c2c55cfd67e02ee9ea105f7e505a8b755cb8273f2f60f3f876163 | 11
 sha512:e98b94787cc5e79f66d225411301c4fa4fcf6b4a48334691eb7dd10be48d2e364b7eec0b37d7f9e65d233e7047ce728540b528b1b52475d5c341e70e78fee33c | 12
(3 rows)

# SELECT COUNT(*) FROM indexreport;
 count 
-------
     3
(1 row)
```
You'll find the expirations for the manifests are within 2 minutes of each other, as expected after messing with the env vars.

### CODE REVIEW

File: scanner/indexer/indexer_test.go
Comment (RTann) [inline scanner/indexer/indexer_test.go:125]: keeping this in case we bump to go1.22 before I merge

File: scanner/indexer/manifest/gc.go
Comment (RTann) [inline scanner/indexer/manifest/gc.go:98]: TODO: move this out of the GC, as this doesn't really make sense here. This doesn't GC, it populates

File: scanner/datastore/postgres/manifest_metadata.go
Comment (dcaravel) [inline scanner/datastore/postgres/manifest_metadata.go:52]: nit: the same is also logged in the caller `scanner/indexer/manifest/manager.go` - consider removing this log or the other if appropriate to do so.

File: scanner/datastore/postgres/manifest_metadata.go
Comment (dcaravel) [inline scanner/datastore/postgres/manifest_metadata.go:22]: super nit:
```suggestion
	// insertMissingManifests inserts missing manifests from the manifest table into manifest_metadata,
	// and it sets the expiration time to some random time in the future between seven and thirty days from now.
```

**Review by dcaravel — COMMENTED** [review]
Is there or should there be a way to disable this feature?

**Comment by RTann** [comment]
> Is there or should there be a way to disable this feature?

@dcaravel no, but I can create a feature flag. That being said, this should definitely be turned on, by default, as we never re-index images nor delete unused ones, at this time.

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go:56]: I considered that, but I wanted to match the indexer migration's time range. Then again, they don't *always* have to match, and that 7 - 30 window from the migration will only hold for the initial migration upon upgrade to this version. I'll look into it more. I don't think I can easily make that initial time range configurable due to how it's implemented, but I can make the subsequent ranges configurable for sure

File: scanner/datastore/postgres/indexer_metadata_store.go
Comment (RTann) [reply scanner/datastore/postgres/indexer_metadata_store.go:1]: I wanted to mimic Claircore as closely as possible, so it'd be easy to understand how they relate - https://github.com/quay/claircore/tree/main/datastore/postgres

File: scanner/indexer/indexer.go
Comment (dcaravel) [reply scanner/indexer/indexer.go:375]: > I figure if we return a failure, we'll retry at some point, and it'll eventually succeed. What do you think?

You're right this may trigger immediate retries in the current scanner v4 client (and we reprocessing eventually after that).  However the retry will trigger a `GetIndexReport` first, which looks like will successfully pull the index report even if the manifest expiry storage failed. 

By returning the index report even if the storage fails you save a round trip.  If `GetIndexReport` is modified to require the manifest expiry exists prior to returning any index report then I can see it making more sense to error as is or self correct - ie: on retrieval if the index report is found but the manifest expiry is not, then insert an expiry (self correct).

**Comment by dcaravel** [comment]
> @dcaravel no, but I can create a feature flag. That being said, this should definitely be turned on, by default, as we never re-index images nor delete unused ones, at this time.

I agree the feature should be turned on - I was more so thinking from the scenario where something goes wrong - worse case we have users unable to scan images using scanner v4 until the next patch - or they can use the 'kill switch' to retain previous behavior while waiting for the patch.

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go]: I addressed it in https://github.com/stackrox/stackrox/pull/12648/commits/3ae55f519f5b0ce541d8ade274f48202be61b9fa

**Review by BradLugo — APPROVED** [review]
Admittedly, a bit of a rubber stamp, but we need to move on this pretty quickly (I don't see anything functionality incorrect - just had some questions and nits)

**Review by jvdm — APPROVED** [review]
Couple of question, no blockers, so I am approving if decide to move forward without blocking on my comments.

File: scanner/datastore/postgres/manifest_metadata.go
Comment (jvdm) [inline scanner/datastore/postgres/manifest_metadata.go:138]: Is there any other metadata related to the IR in CC that we care?

File: scanner/datastore/postgres/manifest_metadata.go
Comment (jvdm) [inline scanner/datastore/postgres/manifest_metadata.go:102]: I am wondering if a big transaction to delete multiple manifests in a single shot might become a point of contention in the future, as likely they will be created close to each other in time.

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go]: I'll lower it to one hour. You're right, it is to protect users, but if they are messing with this variable, they (hopefully) have a good reason

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go]: `MigrateManifests` uses a `TryLock`, so it won't block other replicas. The query may potentially look at a bunch of rows, but that may be ok.

```
WHERE NOT EXISTS (
	SELECT FROM manifest_metadata mm WHERE mm.manifest_id = m.hash
)
```

I'm not a DB expert, but I imagine the slowest part of the query would be this part, which I think may be decently fast, as we are comparing primary keys, which (I believe) are indexed, by default.

So, I think this is ok to do upon startup. Also, from what I can tell, there are no timeouts (I guess this is both good and bad since we don't have to worry about it taking a while, but we have nothing stopping it from halting forever).

I'm ok with keeping this here for now and creating a ticket to consider moving this outside the initialization step

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go:249]: I think we should keep that one, though, because it'll help tell us something is wrong in a way that is harder to ignore than a log. At the moment, anything not migrated to the metadata table is never GC'd, and we shouldn't miss those (there is a ticket to address that). I think I'll keep that error for now, but I can add a comment saying we can consider changing it

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go]: [032edec](https://github.com/stackrox/stackrox/pull/12648/commits/032edec4b65bcc8d1852bd3b4ba09aaa5feb940d)

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go]: Linked a Jira ticket: [032edec](https://github.com/stackrox/stackrox/pull/12648/commits/032edec4b65bcc8d1852bd3b4ba09aaa5feb940d)

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go:249]: Linked a Jira ticket: [032edec](https://github.com/stackrox/stackrox/pull/12648/commits/032edec4b65bcc8d1852bd3b4ba09aaa5feb940d)

File: CHANGELOG.md
Comment (RTann) [inline CHANGELOG.md:88]: Added this unrelated entry, too, because I wanted to add it anyway. Just easier to get it in here vs a separate PR

File: scanner/datastore/postgres/manifest_metadata.go
Comment (RTann) [reply scanner/datastore/postgres/manifest_metadata.go:102]: > as likely the manifests and metadata will be created close to each other in time.

Can you clarify this part? How would this add to any kind of contention?

Comment (RTann) [reply scanner/datastore/postgres/manifest_metadata.go:102]: I can imagine deleting A TON at once could be problematic, but that's why I set the default expiration dates to be some time between 7 days and 30 days from now with the GC checking every 6 hours. Ideally, we have even distribution. Let's say there are 10,000 manifests, then each GC cycle will have to clean up ~110 manifests per cycle (ie about 1.1% of the manifests). I think that's ok, but we should have some kind of metrics to tell us in the future

File: scanner/datastore/postgres/manifest_metadata.go
Comment (RTann) [reply scanner/datastore/postgres/manifest_metadata.go:138]: This function starts its own transaction for each given `claircore.Digest` and deletes all related data (manifest table, indexreport table, etc). This is what's called when we'd want to call [`libindex.DeleteManifests`](https://github.com/quay/claircore/blob/v1.5.33/libindex/libindex.go#L287)

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go:516]: great catch. Agreed, this is ok, as the new Indexers will catch this and then re-index the manifest (essentially just add the entry to the metadata table)

Comment (RTann) [reply scanner/indexer/indexer.go]: Updated the comment: [4fb1d8f](https://github.com/stackrox/stackrox/pull/12648/commits/4fb1d8f76ddaa3d5871d5b5fdbe50bc485b0c566)

File: scanner/datastore/postgres/manifest_metadata.go
Comment (jvdm) [reply scanner/datastore/postgres/manifest_metadata.go:102]: > as likely the manifests and metadata will be created close to each other in time.

Thinking IRs would be likely to be created with a timestamp close to each other; as in when you first install, or when new deployments are done among all services in a cluster, etc. It would mean (hypothetically) that when GC runs likely these IRs would be deleted together, rather than evenly distributed as you described.

Comment (RTann) [reply scanner/datastore/postgres/manifest_metadata.go:102]: gotcha.

So upon upgrade to this version, all manifests are given the same timestamp of "now". Originally it was hardcoded to be a random time between 7 and 30 days from "now" (for each entry. As in, the expiration time was chosen for each manifest instead of being the same for all of them). Due to the known issues with 4.5.4, I switched it to just GC them all immediately (the GC will run upon startup, detect all manifests have expired, and clean them)

From here, let's pretend we get a bunch of scan requests at once. We set the manifests' expiration dates to be some time between 7 and 30 days from "now". This is per-manifest. So, we'll have a bunch of manifests which expire within the same 7 - 30 day window, but this is a rather large window.

I realize the math I presented above isn't perfect because this 7 - 30 days window has a bunch of overlap. As in, let's say 100 manifests expired on day 3. They are now eligible to be re-indexed on days 14 - 37, which overlaps with that original window between days 14 - 30. I'm still tempted to think this is ok, but not as good as I was originally thinking. I'll create a ticket to consider looking into the current re-indexing windows

Comment (RTann) [reply scanner/datastore/postgres/manifest_metadata.go:102]: I redid the math, and each GC cycle would be responsible for about 1,773 manifests starting on day 30 (assuming 10,000 manifests starting on day 1). I can increase the GC frequency to every 4 hours to improve this to be about 1,182 manifests per cycle. I do fear doing the GC any more frequently may be a bit overkill (keeping in mind not every manifest *needs* to be re-indexed every 7 - 30 days, so overdoing it would be wasteful, and add even more pressure to registries). I think this is ok for now. 10,000 images seems like a lot for a cluster. I'll ask around to see what's a reasonable maximum number of images to consider

**Review by dcaravel — CHANGES_REQUESTED** [review]
Can ignore the nits for now given the time sensitivity - but the index name and the feature flag may need to be double checked.

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go:240]: I'll create a Jira to cover this

File: scanner/indexer/manifest/manager.go
Comment (RTann) [reply scanner/indexer/manifest/manager.go:115]: going to change the way I handle these contexts

File: scanner/datastore/postgres/indexer_metadata_store.go
Comment (RTann) [reply scanner/datastore/postgres/indexer_metadata_store.go:52]: true, but it's there in Claircore, so I'd rather keep it for consistency between the two

File: scanner/datastore/postgres/migrations/migrations.go
Comment (RTann) [reply scanner/datastore/postgres/migrations/migrations.go:23]: Good catch. I updated `indexer.go` to check tot he feature flag, too: [fcb08e5](https://github.com/stackrox/stackrox/pull/12648/commits/fcb08e53ac4f52f0f9cadc512c706bcfcb8c53b6)

File: scanner/indexer/manifest/manager.go
Comment (RTann) [reply scanner/indexer/manifest/manager.go]: [fcb08e5](https://github.com/stackrox/stackrox/pull/12648/commits/fcb08e53ac4f52f0f9cadc512c706bcfcb8c53b6)

File: scanner/indexer/indexer.go
Comment (RTann) [reply scanner/indexer/indexer.go]: Added a TODO with the Jira here: [fcb08e5](https://github.com/stackrox/stackrox/pull/12648/commits/fcb08e53ac4f52f0f9cadc512c706bcfcb8c53b6)

**Review by jvdm — APPROVED** [review]

File: scanner/datastore/postgres/manifest_metadata.go
Comment (BradLugo) [inline scanner/datastore/postgres/manifest_metadata.go:106]: Does `Manager` need a locker if all the work is done in a transaction?

**Pull request merged**
